chart/infra-server/static/flavors.yaml

```diff
@@ -344,6 +344,13 @@
       value: false
       kind: optional
 
+    - name: keep-failed-cluster
```

Comment:
I am asking myself if the individual users need to care about this setting or if it would be better to have this as a global control (as in: the users can't select it, we need to switch on this globally in the Helm values and deploy a test environment).
My fear is that we will just keep a lot of failed clusters around in case of an incident like on Monday.

Reply:
> I am asking myself if the individual users need to care about this setting or if it would be better to have this as a global control (as in: the users can't select it, we need to switch on this globally in the Helm values and deploy a test environment).

I don't have a _strong_ opinion on this - I'm down for either. In an ideal world, I think I'd like to have some admin-specific settings so we could decide which individual cluster we'd want to keep. But the option currently implemented in this PR seemed like the most straightforward.

> My fear is that we will just keep a lot of failed clusters around in case of an incident like on Monday.

If I correctly understand how the workflow and automation-flavor work, I think the clusters marked with the `keep-failed-cluster` would still be destroyed at the end of their lifespan, i.e., the destroy step in the create script would be skipped if the create step fails, but the destroy step would still run at the end of the lifespan. I'm not 100% sure on that though, so please correct me if I'm wrong.

That being said, it wouldn't stop folks from seeing the error the first time, reproducing it with the `keep-failed-cluster` flag enabled thinking they're helping us by giving us a cluster that failed. That would indeed be helpful if only one or two people did that, but if everyone started doing that, it would cause even more problems. Thinking out loud - maybe adding some explanation in the help message would be sufficient for that case? Or we could build in a limit to the number of currently provisioned clusters? 
> Also it is just another setting that they need to worry about when spinning up a new cluster.

Agreed. I'm hoping the helping message is descriptive enough to indicate they don't need to worry about it, but yeah, it adds another option in the UI that they'd need to parse.

---

Anyway, I'm totally down to scrap this approach if we want to go in a different direction.

Reply:
> I think I'd like to have some admin-specific settings so we could decide which individual cluster we'd want to keep.

You're right. A global approach (my suggestion) would keep all clusters of a flavor around, while a flag (your approach) would allow for a more specific investigation. 

> If I correctly understand how the workflow and automation-flavor work, I think the clusters marked with the keep-failed-cluster would still be destroyed at the end of their lifespan, i.e., the destroy step in the create script would be skipped if the create step fails, but the destroy step would still run at the end of the lifespan. I'm not 100% sure on that though, so please correct me if I'm wrong.

From what I observed in our debugging of the OCP issues this week, that is correct.

> That would indeed be helpful if only one or two people did that, but if everyone started doing that, it would cause even more problems. Thinking out loud - maybe adding some explanation in the help message would be sufficient for that case? Or we could build in a limit to the number of currently provisioned clusters?

I think this might have become an issue on Monday, when people started to try to provision clusters on their own, just to "check if it works for them". It didn't cause any issues, because failed clusters were immediately destroyed - but IMHO this could have been prevented by better communication (on my side - I chose not to intervene, because we were already too deep in the investigation).

Reply:
@davdhacs @gavin-stackrox what do you think about this?

Reply:
Let's do it. @BradLugo could you document this discussion in the Jira issue?

A follow-up to enable "keeping failed clusters globally" could be to set the default value with Helm.

Reply:
Documented here: https://issues.redhat.com/browse/ROX-16430?focusedId=23856033&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-23856033

Please feel free to add anything I missed.